### PR TITLE
ir: Use the mangling hack in win32 too.

### PR DESCRIPTION
**Pull request by emilio**

### CODE REVIEW

**Comment by emilio** [comment]
r? @fitzgen 

 * Xidorn checked the arch bits, but was going to sleep, so asked me to redirect the rest of the review to you :)

**Comment by fitzgen** [comment]
```
error: missing documentation for a method
  --> src/ir/function.rs:92:5
   |
92 |       pub fn is_unknown(&self) -> bool {
   |  _____^ starting here...
93 | |         match *self {
94 | |             Abi::Unknown(..) => true,
95 | |             _ => false,
96 | |         }
97 | |     }
   | |_____^ ...ending here
   |
note: lint level defined here
  --> src/lib.rs:8:9
   |
8  | #![deny(missing_docs)]
   |         ^^^^^^^^^^^^
error: aborting due to previous error
```

File: src/codegen/mod.rs [hunk 1]
```diff
@@ -2949,8 +2957,15 @@ impl CodeGenerator for Function {
             vis: ast::Visibility::Public,
         };
 
-        let item = ForeignModBuilder::new(signature.abi()
-                .expect("Invalid or unknown ABI for function!"))
+        let abi = match signature.abi() {
+            Abi::Known(abi) => abi,
+            Abi::Unknown(unknown_abi) => {
+                panic!("Invalid or unknown abi {:?} for function {:?} ({:?})",
+                       unknown_abi, canonical_name, self);
```
Comment (fitzgen) [inline src/codegen/mod.rs]: This match is duplicated verbatim: let's pull it out to a `unwrap_known_abi` function.

File: src/ir/context.rs [hunk 2]
```diff
@@ -232,6 +235,32 @@ impl<'ctx> BindgenContext<'ctx> {
                                           parse_options)
                 .expect("TranslationUnit::parse failed");
 
+        // TODO(emilio): Use the CXTargetInfo here when available.
+        //
+        // see: https://reviews.llvm.org/D32389
+        let mut effective_target = None;
+        for opt in &options.clang_args {
+            if opt.starts_with("--target=") {
```
Comment (fitzgen) [inline src/ir/context.rs]: What if the clang args are of the form `--target <blah>` rather than `--target=<blah>`?

File: src/ir/context.rs [hunk 3]
```diff
@@ -232,6 +235,32 @@ impl<'ctx> BindgenContext<'ctx> {
...
+            // If we're running from a build script, try to find the cargo
+            // target.
+            effective_target = env::var("TARGET").ok();
+        }
+
+        if effective_target.is_none() {
+            const HOST_TARGET: &'static str =
+                include_str!(concat!(env!("OUT_DIR"), "/host-target.txt"));
+            effective_target = Some(HOST_TARGET.to_owned());
+        }
```
Comment (fitzgen) [inline src/ir/context.rs]: This could be a chain of `.or_else`s, if you prefer.

File: src/ir/function.rs [hunk 4]
```diff
@@ -79,6 +79,24 @@ impl DotAttributes for Function {
...
+/// An ABI extracted from a clang cursor.
+#[derive(Debug, Copy, Clone)]
+pub enum Abi {
+    /// A known ABI, that rust also understand.
+    Known(abi::Abi),
+    /// An unknown or invalid ABI.
+    Unknown(CXCallingConv),
+}
+
+impl Abi {
```
Comment (fitzgen) [inline src/ir/function.rs]: Aha! Perfect place to add `unwrap_known_abi` ;)

File: src/codegen/mod.rs (hunk 1 above)
Comment (emilio) [inline src/codegen/mod.rs]: It is not, one prints canonical name and a `Function` type. The other doesn't. I expected to add an `unwrap_known_abi` method, but that wouldn't address #727

**Pull request merged**